http-client/src/main/java/io/airlift/http/client/GatheringByteArrayInputStream.java

```diff
@@ -0,0 +1,99 @@
+package io.airlift.http.client;
+
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkElementIndex;
+import static com.google.common.base.Preconditions.checkPositionIndex;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.lang.System.arraycopy;
+import static java.util.Objects.requireNonNull;
+
+public class GatheringByteArrayInputStream
+        extends InputStream
+{
+    private final Iterator<byte[]> buffers;
+    private final byte[] singleByte;
+    private byte[] currentBuffer;
+    private int currentBufferPosition;
+    private long remainingBytes;
+
+    public GatheringByteArrayInputStream(List<byte[]> buffers, long totalBytes)
+    {
+        checkArgument(totalBytes >= 0, "totalBytes should equal to or greater than 0");
+        this.buffers = requireNonNull(buffers, "buffers is null").iterator();
+        this.remainingBytes = totalBytes;
+
+        currentBuffer = new byte[0];
+        currentBufferPosition = 0;
```

Comment:
don't initialize a field to the java default unless required because it is `final`